internal/services/pam/pam_test.go

```diff
@@ -0,0 +1,596 @@
+
+		"Empty data gets JSON formatted": {username: "IA_empty_data"},
+
+		// service errors
+		"Error when sessionID is empty": {sessionID: "-"},
+		"Error when there is no broker": {sessionID: "no broker", brokerID: "does not exist"},
+
+		// broker errors
+		"Error when authorizing":                            {username: "IA_error"},
+		"Error when broker returns invalid access":          {username: "IA_invalid"},
+		"Error when broker returns invalid data":            {username: "IA_invalid_data"},
+		"Error when calling second time without cancelling": {username: "IA_second_call", secondCall: true},
+	}
+	for name, tc := range tests {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			if tc.brokerID == "" {
+				tc.brokerID = generatedID
+			}
+
+			if tc.brokerID != "does not exist" {
+				if tc.username == "" {
+					tc.username = "does not matter"
+				}
+				// Since we use the same broker here, we need to prefix the username to prevent test cases from conflicting.
+				tc.username = t.Name() + testutils.IDPrefix + tc.username
+				sbResp, err := testClient.SelectBroker(context.Background(), &authd.SBRequest{
+					BrokerId: tc.brokerID,
+					Username: tc.username,
+				})
+				require.NoError(t, err, "Setup: failed to create session for tests")
+				if tc.sessionID == "" {
+					tc.sessionID = sbResp.GetSessionId()
+				}
+			}
+
+			if tc.sessionID == "-" {
+				tc.sessionID = ""
+			}
+
+			var firstCall, secondCall string
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			done := make(chan struct{})
+			go func() {
+				defer close(done)
+				iaReq := &authd.IARequest{
+					SessionId:          tc.sessionID,
+					AuthenticationData: "some data",
+				}
+				iaResp, err := testClient.IsAuthorized(ctx, iaReq)
+				firstCall = fmt.Sprintf("FIRST CALL:\n\taccess: %s\n\tdata: %s\n\terr: %v\n",
+					iaResp.GetAccess(),
+					iaResp.GetData(),
+					err,
+				)
+			}()
+			// Give some time for the first call to block
+			time.Sleep(time.Second)
+			if tc.cancelFirstCall {
+				cancel()
+				time.Sleep(time.Millisecond)
+				<-done
+			}
+
+			if tc.secondCall {
+				iaReq := &authd.IARequest{
+					SessionId:          tc.sessionID,
+					AuthenticationData: "some data",
+				}
+				iaResp, err := testClient.IsAuthorized(context.Background(), iaReq)
+				secondCall = fmt.Sprintf("SECOND CALL:\n\taccess: %s\n\tdata: %s\n\terr: %v\n",
+					iaResp.GetAccess(),
+					iaResp.GetData(),
+					err,
+				)
+			}
+			<-done
+			got := firstCall + secondCall
+			want := testutils.LoadWithUpdateFromGolden(t, got)
+			require.Equal(t, want, got, "IsAuthorized should return the expected combined data, but did not")
+		})
+	}
+}
+
+func TestEndSession(t *testing.T) {
+	t.Parallel()
+
+	tests := map[string]struct {
+		username  string
+		brokerID  string
+		sessionID string
+
+		wantErr bool
+	}{
+		"Successfully end session": {username: "success"},
```

Comment:
here too, it’s dangerous to user "success" because you can have this username reused in multiple parallel tests. It should really be test base with the transformation you did like in the other one by prefixing it.